### ISSUE TITLE
OMEdit: inconsistent handling of protected objects

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#4346]
[Ticket#4346](https://trac.openmodelica.org/OpenModelica/ticket/4346)

**Comment by sbaetzing** [comment]
In that issue it would be nice if protected / hided components / variables((hideResult =true) ) are not displayed in the variable section in Plotting tab. since they are not part of the result file and not available after clicking on them.

**Comment by adeas31** [comment]
The MWE provided in [Ticket#4346](https://trac.openmodelica.org/OpenModelica/ticket/4346) is bit wrong. Here is the updated MWE,

```
model test
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 1)  annotation(
    Placement(visible = true, transformation(origin = {0, 20}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
protected
  Real test = 123;
  Modelica.Electrical.Analog.Basic.Ground ground annotation(HideResult=true,
    Placement(visible = true, transformation(origin = {0, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(resistor.n, ground.p) annotation(
    Line(points = {{0, 10}, {0, 10}, {0, 0}, {0, 0}}, color = {0, 0, 255}));
  annotation(
    uses(Modelica(version = "3.2.2")));
end test;
```

Anything marked `protected` gets `hideResult = "true"` in the `model_init.xml` file. If the we remove `protected` then `ground` should still get `hideResult = "true"` due to `HideResult` annotation but that is not the case.

```
model test
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 1)  annotation(
    Placement(visible = true, transformation(origin = {0, 20}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
  Real test = 123;
  Modelica.Electrical.Analog.Basic.Ground ground annotation(HideResult=true,
    Placement(visible = true, transformation(origin = {0, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(resistor.n, ground.p) annotation(
    Line(points = {{0, 10}, {0, 10}, {0, 0}, {0, 0}}, color = {0, 0, 255}));
  annotation(
    uses(Modelica(version = "3.2.2")));
end test;
```

We handle the `HideResult` annotation only on primitive datatypes,

```
model test
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 1)  annotation(
    Placement(visible = true, transformation(origin = {0, 20}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
  Real test = 123 annotation(HideResult=true);
  Modelica.Electrical.Analog.Basic.Ground ground annotation(HideResult=true,
    Placement(visible = true, transformation(origin = {0, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(resistor.n, ground.p) annotation(
    Line(points = {{0, 10}, {0, 10}, {0, 0}, {0, 0}}, color = {0, 0, 255}));
  annotation(
    uses(Modelica(version = "3.2.2")));
end test;
```

**Comment by casella** [comment]
@sbaetzing, is the behaviour in the current nighly build acceptable for you?

**Comment by casella** [comment]
@niklwors, @sbaetzing, can you please check the behaviour of the current nightly build?

**Comment by sbaetzing** [comment]
@casella, i am very sorry for the late response. 
Tested it again with nightly build (27.11.2022). Behavior seems to be identical.

value of hideResult annotation is still only evaluated at parameter and variables, not class.
Also it is still not possible to set its value depending on a other parameter.

best regards,

Sven

Note: For us it has more benefits to evaluate the annotation at classes than setting its value via parameter.

**Comment by casella** [comment]
@adeas31 should the NF propagate the annotation from the instantiated object all the way down to individual variable declarations, so they end up in the xml file?

**Comment by adeas31** [comment]
I think the annotations are propagated since it works for the primitive types.

**Comment by casella** [comment]
@adeas31, I'm not sure. If I take the `test` model shown above
```modelica
model test
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 1)  annotation(
    Placement(visible = true, transformation(origin = {0, 20}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
  Real test = 123 annotation(HideResult=true);
  Modelica.Electrical.Analog.Basic.Ground ground annotation(HideResult=true,
    Placement(visible = true, transformation(origin = {0, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(resistor.n, ground.p) annotation(
    Line(points = {{0, 10}, {0, 10}, {0, 0}, {0, 0}}, color = {0, 0, 255}));
  annotation(
    uses(Modelica(version = "3.2.2")));
end test;
```
when I flatten it I get
```
class test
  parameter Real resistor.R(quantity = "Resistance", unit = "Ohm", start = 1.0) = 1.0 "Resistance at temperature T_ref";
  parameter Real resistor.T_ref(quantity = "ThermodynamicTemperature", unit = "K", displayUnit = "degC", min = 0.0, start = 288.15, nominal = 300.0) = 300.15 "Reference temperature";
  parameter Real resistor.alpha(quantity = "LinearTemperatureCoefficient", unit = "1/K") = 0.0 "Temperature coefficient of resistance (R_actual = R*(1 + alpha*(T_heatPort - T_ref))";
  Real resistor.v(quantity = "ElectricPotential", unit = "V") "Voltage drop of the two pins (= p.v - n.v)";
  Real resistor.p.v(quantity = "ElectricPotential", unit = "V") "Potential at the pin";
  Real resistor.p.i(quantity = "ElectricCurrent", unit = "A") "Current flowing into the pin";
  Real resistor.n.v(quantity = "ElectricPotential", unit = "V") "Potential at the pin";
  Real resistor.n.i(quantity = "ElectricCurrent", unit = "A") "Current flowing into the pin";
  Real resistor.i(quantity = "ElectricCurrent", unit = "A") "Current flowing from pin p to pin n";
  final parameter Boolean resistor.useHeatPort = false "= true, if heatPort is enabled" annotation(Evaluate = true);
  parameter Real resistor.T(quantity = "ThermodynamicTemperature", unit = "K", displayUnit = "degC", min = 0.0, start = 288.15, nominal = 300.0) = resistor.T_ref "Fixed device temperature if useHeatPort = false";
  Real resistor.LossPower(quantity = "Power", unit = "W") "Loss power leaving component via heatPort";
  Real resistor.T_heatPort(quantity = "ThermodynamicTemperature", unit = "K", displayUnit = "degC", min = 0.0, start = 288.15, nominal = 300.0) "Temperature of heatPort";
  Real resistor.R_actual(quantity = "Resistance", unit = "Ohm") "Actual resistance = R*(1 + alpha*(T_heatPort - T_ref))";
  Real test = 123.0;
  Real ground.p.v(quantity = "ElectricPotential", unit = "V") "Potential at the pin";
  Real ground.p.i(quantity = "ElectricCurrent", unit = "A") "Current flowing into the pin";
equation
  resistor.n.v = ground.p.v;
  resistor.p.i = 0.0;
  ground.p.i + resistor.n.i = 0.0;
  assert(1.0 + resistor.alpha * (resistor.T_heatPort - resistor.T_ref) >= 1e-15, "Temperature outside scope of model!");
  resistor.R_actual = resistor.R * (1.0 + resistor.alpha * (resistor.T_heatPort - resistor.T_ref));
  resistor.v = resistor.R_actual * resistor.i;
  resistor.LossPower = resistor.v * resistor.i;
  resistor.T_heatPort = resistor.T;
  0.0 = resistor.p.i + resistor.n.i;
  resistor.i = resistor.p.i;
  resistor.v = resistor.p.v - resistor.n.v;
  ground.p.v = 0.0;
end test;
```
The two declarations of the `ground` variables (`ground.p.v` and `ground.p.i`) have no annotation. This may be simply because the instantiateModel API function strips all the annotations away, so I can't see them because of that. If you check the AST, is the `HideResult` annotation propagated to those variables?

**Comment by adeas31** [comment]
Is there in the AST for sure.

The following script,

```
loadString("model test
  Modelica.Electrical.Analog.Basic.Resistor resistor(R = 1)  annotation(
    Placement(visible = true, transformation(origin = {0, 20}, extent = {{-10, -10}, {10, 10}}, rotation = -90)));
  Real test = 123 annotation(HideResult=true);
  Modelica.Electrical.Analog.Basic.Ground ground annotation(HideResult=true,
    Placement(visible = true, transformation(origin = {0, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(resistor.n, ground.p) annotation(
    Line(points = {{0, 10}, {0, 10}, {0, 0}, {0, 0}}, color = {0, 0, 255}));
  annotation(
    uses(Modelica(version = \"3.2.2\")));
end test;");
getErrorString();
getComponentAnnotations(test);
getErrorString();
```
produces,
```
true
""
{{Placement(true,0.0,20.0,-10.0,-10.0,10.0,10.0,-90.0,-,-,-,-,-,-,)},{HideResult=true},{HideResult=true, Placement(true,0.0,-10.0,-10.0,-10.0,10.0,10.0,0.0,-,-,-,-,-,-,)}}
""
```

`getComponentAnnotations` reads from AST.

**Comment by perost** [comment]
The frontend does not propagate any annotations downward onto variables, so annotations on a class instance will not be part of the flattened model since only the indidual variables end up there. The annotations are available in the new instance API though, since it's hierarchical rather than flat, so if we use that then OMEdit should be able to keep track of which variables should be hidden because their parent is hidden.

**Comment by adeas31** [comment]
For this the new instance api is not used. The problem here is that we only add `hideResult = "true"` for primitive types in model_init.xml file. So `ground` end up in the xml file without `hideResult = "true"`.

**Comment by casella** [comment]
I guess some kind of top-down propagation of the hideResult annotation will have to be put in place, one way or the other.

**Comment by kabdelhak** [comment]
I added the attribute propagation in #9859. It seems to work fine since the sim code (`-d=dumpSimCode`) shows the variables as follows:
```
modelInfo:
========================================
algVars (7)
----------------------
index:0: ground.p.v (no alias)  protected  hideResult  initial: 	no arrCref index:(1) []
index:1: resistor.LossPower (no alias)  initial: 	no arrCref index:(2) []
index:2: resistor.R_actual (no alias)  initial: 	no arrCref index:(3) []
index:3: resistor.n.i (no alias)  initial: 	no arrCref index:(4) []
index:4: resistor.p.i (no alias)  initial: 	no arrCref index:(5) []
index:5: resistor.v (no alias)  initial: 	no arrCref index:(6) []
index:6: test (no alias)  protected  initial: 	no arrCref index:(7) []
aliasVars
----------------------
index:0: ground.p.i (alias: resistor.p.i)  protected  hideResult  initial: 	no arrCref index:(13) []
index:1: resistor.i (alias: resistor.p.i)  initial: 	no arrCref index:(14) []
index:2: resistor.n.v (alias: ground.p.v)  initial: 	no arrCref index:(15) []
index:3: resistor.p.v (alias: resistor.v)  initial: 	no arrCref index:(16) []
paramVars
----------------------
index:0: resistor.R (no alias)  initial: 1.0	no arrCref index:(8) []
index:1: resistor.T (no alias)  initial: resistor.T_ref	no arrCref index:(9) []
index:2: resistor.T_heatPort (no alias)  initial: resistor.T	no arrCref index:(10) []
index:3: resistor.T_ref (no alias)  initial: 300.15	no arrCref index:(11) []
index:4: resistor.alpha (no alias)  initial: 0.0	no arrCref index:(12) []
boolParamVars
----------------------
index:0: resistor.useHeatPort (no alias)  hideResult  initial: false	no arrCref index:(17) []
```
As we can see `ground.p.v ` and ` ground.p.i` have the `hideResult` attribute. When observing the result file though, `ground.p.v ` is still present. My assumption is, that it still has to be there because of
```
index:2: resistor.n.v (alias: ground.p.v)  initial: 	no arrCref index:(15) []
```
We cannot eliminate it from the result file since it has to be used to show this alias variable. I suppose this is expected behaviour, one would need to take `hideResult` in consideration when choosing alias variables.

Nevertheless my PR seems to be an upgrade to whats going on, i will merge it if the testsuite and @perost agree with my approach.

@casella Any thoughts about the alias thing?

**Comment by kabdelhak** [comment]
The original issue is resolved with #9859, if there is something to do regarding the alias removal or another model feel free to reopen this ticket or make a new one.

**Comment by casella** [comment]
I guess discussing alias obfuscation is a bit of hair-splitting, but we can do that in another ticket if there is a specific requirement about that.
